### PR TITLE
AlmaLinux/build-system#74: Update VM driver to support testing in VMs  [part of VMs orchestrating]; AlmaLinux/build-system#91: Update VM driver to support testing in VMs [part of tests]

### PR DESCRIPTION
**Pull request by soksanichenko**
- Path refactoring
- The separate section of ALTS config for OpenNebula options
- Improved method for searching image ID/template ID in OpenNebula
- Verbose output for binary ansible-playbook
- Method for rebooting a target VM

### CODE REVIEW

**Review by anfimovdm — CHANGES_REQUESTED** [review]
tests are broken

File: alts/worker/tasks.py [hunk 1]
```diff
@@ -95,11 +95,11 @@ def is_success(stage_data_: dict):
...
     runner_kwargs = {
         'repositories': task_params.get('repositories')
-                        if task_params.get('repositories') else [],
+        if task_params.get('repositories') else [],
         'dist_arch': task_params.get('dist_arch')
-                     if task_params.get('dist_arch') else 'x86_64',
+        if task_params.get('dist_arch') else 'x86_64',
         'test_configuration': task_params.get('test_configuration')
-                              if task_params.get('test_configuration') else {},
+        if task_params.get('test_configuration') else {},
```
Comment (anfimovdm) [inline alts/worker/tasks.py:102]: I think we can do the following:
```python
runner_kwargs = {
    'repositories': task_params.get('repositories') or [],
    'dist_arch': task_params.get('dist_arch') or 'x86_64',
    ...
}
```

File: resources/opennebula/opennebula.tf.tmpl [hunk 2]
```diff
@@ -24,9 +24,8 @@ resource "opennebula_virtual_machine" "${vm_name}" {
     model = "host-passthrough"
   }
   disk {
-    image_id: ${image_id}
-    size: ${disk_size}
```
Comment (anfimovdm) [inline resources/opennebula/opennebula.tf.tmpl:28]: Why don't we need the `size` parameter anymore?

**Review by anfimovdm — CHANGES_REQUESTED** [review]

File: resources/opennebula/opennebula.tf.tmpl (hunk 2 above)
Comment (soksanichenko) [reply resources/opennebula/opennebula.tf.tmpl:28]: Accepted

File: alts/worker/tasks.py (hunk 1 above)
Comment (soksanichenko) [reply alts/worker/tasks.py:102]: Accepted

File: alts/shared/utils/file_utils.py [hunk 3]
```diff
@@ -49,8 +50,8 @@ def feed_hasher(_fd):
                 buff = buff.encode('utf')
             hasher.update(buff)
             buff = _fd.read(buff_size)
-    if isinstance(file_path, str):
-        with open(file_path, "rb") as fd:
+    if isinstance(file_path, Path):
+        with file_path.open(mode='rb') as fd:
             feed_hasher(fd)
```
[the remark this answers is not in the document]
Comment (soksanichenko) [reply alts/shared/utils/file_utils.py:55]: I forgot to rollback all changes about Path

File: alts/shared/uploaders/pulp.py [hunk 4]
```diff
@@ -220,7 +223,11 @@ def upload(self, artifacts_dir: str, **kwargs) -> List[dict]:
         self._logger.info('Starting files upload')
         with ThreadPoolExecutor(max_workers=self._concurrency) as executor:
             futures = {
-                executor.submit(self.upload_single_file, artifact): artifact
+                executor.submit(
+                    self.upload_single_file,
+                    artifact,
+                    'test_log',
```
[the remark this answers is not in the document]
Comment (soksanichenko) [reply alts/shared/uploaders/pulp.py:229]: I know, but the warning from IDE is so annoying
изображение

File: alts/shared/uploaders/azure.py [hunk 5]
```diff
@@ -27,29 +28,32 @@ def __init__(self, connection_string: str, container_name: str):
...
+    ) -> str:
+        blob_name = azure_upload_dir.joinpath(file_path.name)
         blob_client = self._blob_client.get_blob_client(
-            container=self._container_name, blob=blob_name)
+            container=self._container_name, blob=str(blob_name))
         try:
-            with open(file_path, 'rb') as f:
-                blob_client.upload_blob(f)
+            with file_path.open(mode='rb') as fd:
+                blob_client.upload_blob(fd)
```
[the remark this answers is not in the document]
Comment (soksanichenko) [reply alts/shared/uploaders/azure.py:41]: I forgot to rollback all changes about Path

File: alts/shared/uploaders/azure.py [hunk 6]
```diff
@@ -67,12 +71,11 @@ def upload(self, artifacts_dir: str, **kwargs) -> \
         if not kwargs.get('upload_dir'):
             self._logger.error(self.argument_required_message)
             raise UploadError(self.argument_required_message)
-        azure_upload_dir = kwargs.get('upload_dir')
+        azure_upload_dir = kwargs.get('upload_dir')  # type: Path
```
[the remark this answers is not in the document]
Comment (soksanichenko) [reply alts/shared/uploaders/azure.py:74]: We can't because the base method has another signature

**Review by anfimovdm — CHANGES_REQUESTED** [review]

File: resources/ansible.cfg [hunk 7]
```diff
@@ -1,2 +1,9 @@
 [defaults]
+stdout_callback=debug
```
Comment (anfimovdm) [inline resources/ansible.cfg:2]: when I'm trying to run tests via docker runner I get the following error
```
ERROR! Invalid callback for stdout specified: debug
```

**Review by anfimovdm — CHANGES_REQUESTED** [review]

**Review by anfimovdm — APPROVED** [review]

**Review by zklevsha — APPROVED** [review]

**Review by maccelf — APPROVED** [review]

**Pull request merged**